### PR TITLE
Server side rendering

### PR DESCRIPTION
**Pull request by KutnerUri**
## Proposed Changes

- Currently runs for every request with `?rendeirng=server` query param.
- Separate bundle for SSR code (in Scope mode)
- Server side rendering at `ui.ui`, with lifecycle hooks to enrich any point of the process.
- Render Lifecycle takes care of rendering, as well as passing state from server to client.
- Migrate `graphql.ui` to use the new slots entirely, instead of being a part of `ui.ui`.
- Prevent any usage of window during server side execution, using `typeof window !== "undefined"`
- Limited solution to GQL ids and caching.

### TO FINISH:
- [x] cleanup, replace console.log()s, etc
- [x] Make sure preview/docs work correctly (threw an error about 'react-dom/server')
- [x] set html title
- [x] middleware should avoid static files. (Maybe move static files to '/public/*' url ?)
- [ ] Harmony instance (make sure it's singleton, and consider making a different instance per request)

### CODE REVIEW

File: scopes/ui-foundation/ui/webpack/webpack.base.config.ts [hunk 1]
```diff
@@ -116,102 +109,22 @@ module.exports = function (workspaceDir, entryFiles, title) {
...
 
       filename: 'static/js/[name].[contenthash:8].js',
       // TODO: remove this when upgrading to webpack 5
       futureEmitAssets: true,
       // There are also additional JS chunk files if you use code splitting.
       chunkFilename: 'static/js/[name].[contenthash:8].chunk.js',
-      // webpack uses `publicPath` to determine where the app is being served from.
-      // It requires a trailing slash, or the file assets will get an incorrect path.
-      // We inferred the "public path" (such as / or /my-project) from homepage.
-      publicPath: `/`,
```
Comment (KutnerUri) [inline scopes/ui-foundation/ui/webpack/webpack.base.config.ts:134]: moved to webpack.browser.config.ts

File: scopes/ui-foundation/ui/webpack/webpack.base.config.ts [hunk 2]
```diff
@@ -116,102 +109,22 @@ module.exports = function (workspaceDir, entryFiles, title) {
...
       // this defaults to 'window', but by setting it to 'this' then
       // module chunks which are built will work in web workers as well.
       globalObject: 'this',
     },
 
-    optimization: {
```
Comment (KutnerUri) [inline scopes/ui-foundation/ui/webpack/webpack.base.config.ts:140]: moved to webpack.browser.config.ts

File: scopes/ui-foundation/ui/webpack/webpack.base.config.ts [hunk 3]
```diff
@@ -426,30 +340,6 @@ module.exports = function (workspaceDir, entryFiles, title) {
       ],
     },
     plugins: [
-      new HtmlWebpackPlugin(
```
Comment (KutnerUri) [inline scopes/ui-foundation/ui/webpack/webpack.base.config.ts:429]: moved to webpack.browser.config.ts

File: scopes/ui-foundation/ui/webpack/webpack.base.config.ts [hunk 4]
```diff
@@ -464,7 +354,6 @@ module.exports = function (workspaceDir, entryFiles, title) {
       //   can be used to reconstruct the HTML if necessary
       new ManifestPlugin({
         fileName: 'asset-manifest.json',
-        publicPath: 'public',
```
Comment (KutnerUri) [inline scopes/ui-foundation/ui/webpack/webpack.base.config.ts:467]: inferred from output.publicPath

**Review by GiladShoham — APPROVED** [review]

**Pull request merged**